### PR TITLE
Bug 1140095 - speed up and animate customizer opening and closing

### PR DESCRIPTION
**Pull request by davidflanagan**

### CODE REVIEW

File: app/js/controllers/main.js [hunk 1]
```diff
@@ -22,7 +22,6 @@ export default class MainController extends Controller {
 
     Gesture.detect(openGesture).then(() => {
       this.attachView();
```
Comment (DouglasSherk) [inline app/js/controllers/main.js]: We could write this as:

``` js
Gesture.detect(openGesture).then(this.attachView.bind(this));
```

File: app/js/controllers/main.js [hunk 2]
```diff
@@ -22,7 +22,6 @@ export default class MainController extends Controller {
...
-      this.waitToBeClosed();
```
Comment (DouglasSherk) [inline app/js/controllers/main.js]: Same as https://github.com/fxos/customizer/pull/9/files#r25951795

File: app/js/controllers/main.js [hunk 3]
```diff
@@ -41,7 +40,6 @@ export default class MainController extends Controller {
 
     Gesture.detect(closeGesture).then(() => {
       this.removeView();
```
Comment (DouglasSherk) [inline app/js/controllers/main.js]: Same as https://github.com/fxos/customizer/pull/9/files#r25951795

File: app/js/controllers/main.js [hunk 4]
```diff
@@ -50,22 +48,34 @@ export default class MainController extends Controller {
...
 
-    var padding = this.view.el.querySelector('fxos-customizer').offsetHeight;
-    this._styleEl = document.createElement('style');
-    this._styleEl.textContent = `
-    html, body {
-      overflow: initial !important;
-    }
+    // And display it. This has to be in a setTimeout or the CSS animation
+    // won't work correctly since we just added it above.
+    setTimeout(() => { this.view.show(); });
```
Comment (DouglasSherk) [inline app/js/controllers/main.js]:
``` js
setTimeout(this.view.show.bind(this));
```

File: app/js/controllers/main.js [hunk 5]
```diff
@@ -50,22 +48,34 @@ export default class MainController extends Controller {
...
-    document.head.appendChild(this._styleEl);
+      body {
+        padding-bottom: ${padding}px !important;
+      }`;
+      document.head.appendChild(this._styleEl);
 
-    this.attached = true;
+      this.attached = true;
+      this.waitToBeClosed(); // We're open, so expect a close gesture
+    }, this.view.animationLength);
```
Comment (DouglasSherk) [inline app/js/controllers/main.js]: There's an 'animationend' event that we could use instead:
https://developer.mozilla.org/en-US/docs/Web/Events/animationend?redirectlocale=en-US&redirectslug=DOM%2FMozilla_event_reference%2Fanimationend

File: app/js/views/main.js [hunk 6]
```diff
@@ -104,6 +104,29 @@ var mainViewTemplate =
 
     --checkbox-border-color: var(--background-minus-minus);
   }
+
+  fxos-customizer {
```
Comment (DouglasSherk) [inline app/js/views/main.js]: This should actually go inside the `::host` selector in `app/js/elements/fxos-customizer.js`. Please move it there instead.

File: app/js/views/main.js [hunk 7]
```diff
@@ -104,6 +104,29 @@ var mainViewTemplate =
...
+    background-color: rgba(0, 0, 0, 0.8);
+    position: fixed;
+    left: 0;
+    right: 0;
+    top: 100%; /* off-screen by default, animated translate to show and hide */
+    height: 50%;
+    border-top: 1px solid #ccc;
+    /*
+     * this needs to go on top of the regular app, but below
+     * gaia-modal and gaia-dialog which we override elsewhere.
```
Comment (DouglasSherk) [inline app/js/views/main.js]: A bit nit-picky, but this is overly-specific. We may add more components or remove the ones here, and it's very likely that we'll forget about this comment.

File: app/js/controllers/main.js [hunk 8]
```diff
@@ -50,22 +48,34 @@ export default class MainController extends Controller {
...
+    setTimeout(() => { this.view.show(); });
+
+    // Wait a bit before populating it and defining the new stylesheet
+    // so that we can just get it on the screen quickly and don't interfere
+    // with the animation
+    setTimeout(() => {
+      this.view.customizer.setRootNode(document.documentElement);
+
+      var padding = this.view.el.querySelector('fxos-customizer').offsetHeight;
+      this._styleEl = document.createElement('style');
```
Comment (DouglasSherk) [inline app/js/controllers/main.js]: This should be `this.view.styleEl` instead, or something like that. I know that I originally wrote this, but in retrospect, I think it's wrong.

File: app/js/controllers/main.js (hunk 4 above)
Comment (davidflanagan) [inline app/js/controllers/main.js]: I've always found this.f.bind(this) to be repetitive and ugly; kind of an anti-pattern. And arrow functions allow us to avoid it and still be concise. Since both create a new function, I assume that they have similar performance characteristics, and I'd prefer to just use => instead of bind in situations like this. I think it makes the code clearer.

Comment (DouglasSherk) [inline app/js/controllers/main.js]: Sure, it's not a strong opinion. I'm fine with that.

File: app/js/controllers/main.js (hunk 5 above)
Comment (davidflanagan) [inline app/js/controllers/main.js]: I know about transitionend but hadn't heard of animationend.  I always worry a little about using those because I don't know what happens if gecko implements some "don't do animations for low-end hardware or for people who get motion sickness" mode and doesn't do the animations. Do we know for sure that the event handlers will still fire?  That's why I went with setTimeout here. But I'll take a look at animationend.

Comment (DouglasSherk) [inline app/js/controllers/main.js]: 'animationend' is reliable, and big parts of Gaia depend on it. I would imagine that, if we ever implemented this no-transition functionality in Gecko, we would still dispatch the 'animationend' event as logic generally doesn't depend on the transition frames actually happening.

Comment (davidflanagan) [inline app/js/controllers/main.js]: Okay, I'll change this. I assume you mean transitionend, though, or does animationend work for both animations and transitions?

Another reason to use setTimeout() would be if we want to start populating the tree during the animation sometime.  I tried a 100ms timeout with a 200ms transition though and it seemed to interfere, so just waiting for the event is probably the right thing in general.

Comment (DouglasSherk) [inline app/js/controllers/main.js]: Yeah, that was the reason I originally wanted to have a dummy Customizer and a real one. We could animate the dummy one while rendering and painting the real one in the background. Then we could just hide the dummy one at the end of the transition. It sounds like that's not worth doing, though.

File: app/js/views/main.js (hunk 6 above)
Comment (davidflanagan) [inline app/js/views/main.js]: You want me to move all of it, even the element positioning, z-index and transition stuff? That would make the customizer component very specific. These seem like view-level details rather that component level detail. Though if the component is app-specific rather than shared, then maybe it makes sense to put really specific CSS in it.

I can see that it might make sense to set the background color in the element, though we should probably use one of the variables defined in this file rather than hard-coding the color.

File: app/js/controllers/main.js (hunk 8 above)
Comment (davidflanagan) [inline app/js/controllers/main.js]: Its weird to have part of a view that affects the entire containing document. But I agree that it doesn't belong in the controller. I'll move it.  I'm considering just adding the stylesheet once and enabling and disabling it (I think we can do that) instead of recreating it each time.  We'd have to hardcode the padding height, but since the view will be hard-coding the position of the customizer, I don't see any problem also hard-coding the padding.

File: app/js/views/main.js (hunk 6 above)
Comment (DouglasSherk) [inline app/js/views/main.js]: Yes, I think all of this should go in the element's `<style scoped>`. I don't think it's unfair to make some assumptions about the use cases for this. The `::host` selector can be overridden by client content if they don't like our decisions. Using CSS vars makes sense, though.

File: app/js/controllers/main.js (hunk 8 above)
Comment (DouglasSherk) [inline app/js/controllers/main.js]: `MainView` seems as good as any place to add stuff like this. It's pretty much the catch-all for elements that don't really fit anywhere else.

File: app/js/controllers/main.js (hunk 4 above)
Comment (justindarc) [inline app/js/controllers/main.js]: I think you can also do:

```
setTimeout(=> this.view.show());
```

(I might be wrong though)

Comment (davidflanagan) [inline app/js/controllers/main.js]: I looked it up. Can't leave out the parens if there are no args, but I can leave out the braces if the body is just a single expression!  Simplest case is for single-arg functions when you can leave out the parens x=>x+1 for example.  But in this case I need ()=>this.view.show().  Good to get rid of braces, though.

File: app/js/app.js [hunk 9]
```diff
@@ -72,3 +72,9 @@ appendChildController.mainController = mainController;
 moveController.mainController = mainController;
 
 window.mainController = mainController;
+
+// We insert the main view into the document on startup.  This adds a
+// small disabled stylesheet and a minimal offscreen element to the
+// DOM. The mainView keeps the rest of the customizer out of the DOM
+// until its open() method is called.
+document.body.appendChild(mainView.el);
```
Comment (DouglasSherk) [inline app/js/app.js]: Why is this needed? We should avoid appending views outside control logic if at all possible. Thus, the ideal flow should be:
`app.js -> RoutingController > MainController > MainView`
So we could append `MainView` inside `MainController`'s `constructor()` or `init()` functions.

File: app/js/controllers/main.js [hunk 10]
```diff
@@ -38,48 +32,23 @@ export default class MainController extends Controller {
...
-    this.attached = true;
+  // These private _wait methods call each other. If you try to call them
+  // directly, you'll end up with more than one gesture detector listening
+  // at a time.
+  _waitToBeOpened() {
+    Gesture.detect(this.openGesture)
+      .then(() => this.view.open())
+      .then(() => {
+        this.view.customizer.setRootNode(document.documentElement);
+        this._waitToBeClosed();
```
Comment (DouglasSherk) [inline app/js/controllers/main.js:52]: Nit-picking: it's weird how this is broken up into two `then()` blocks as the whole thing could really execute as one block, I think. Maybe I'm missing something, though. Is this being done so that we have time for rendering before performing any logic on newly created/appended elements?

File: app/js/controllers/main.js [hunk 11]
```diff
@@ -3,12 +3,11 @@
...
   constructor(options) {
     super(options);
-    this.attached = false;
-    this.waitToBeOpened();
+    this._waitToBeOpened();
   }
 
-  waitToBeOpened() {
-    var openGesture = {
+  get openGesture() {
```
Comment (DouglasSherk) [inline app/js/controllers/main.js:16]: The gesture control code here is big enough that it might be overwhelming `MainController`. That is, instead of `MainController` just tying everything together, it has a significant amount of body logic in it. Maybe we should move this into a new `GestureService` or `GestureController`. This isn't a strong opinion though.

File: app/js/views/main.js [hunk 12]
```diff
@@ -151,14 +204,75 @@ export default class MainView extends View {
     return mainViewTemplate;
   }
 
-  render() {
-    super();
-
-    this.el.appendChild(this.actionMenuView.el);
+  _addChildViews() {
```
Comment (DouglasSherk) [inline app/js/views/main.js:217]: Maybe instead of having to maintain all of these child views in this way, we could have a `div.inner` element or some such which we append/remove as needed. That way, setup and teardown is just a single operation.

File: app/js/controllers/main.js [hunk 13]
```diff
@@ -38,48 +32,23 @@ export default class MainController extends Controller {
...
-
-    body {
-      padding-bottom: ${padding}px !important;
-    }`;
-    document.head.appendChild(this._styleEl);
-
-    this.attached = true;
+  // These private _wait methods call each other. If you try to call them
+  // directly, you'll end up with more than one gesture detector listening
+  // at a time.
```
Comment (DouglasSherk) [inline app/js/controllers/main.js:46]: Seems like a redundant comment to me.

File: app/js/app.js (hunk 9 above)
Comment (davidflanagan) [inline app/js/app.js]: Okay, I'll move that back. I did it this way because in this app, none of the views insert themselves. The main view is in charge of inserting all of the child views, so I figured that I should go up a level from main view to insert the that.  I'll just put it in main controller, then.

File: app/js/controllers/main.js (hunk 10 above)
Comment (davidflanagan) [inline app/js/controllers/main.js:52]: What you're missing is the implicit return for arrow functions that have only one expression.  this.view.open() returns a Promise, which is returned from the then() call. open() does not resolve its promise until the animation is done. In order to keep the animation speed high, it is important that setRootNode() does not get called until that promise resolves.

File: app/js/controllers/main.js (hunk 11 above)
Comment (davidflanagan) [inline app/js/controllers/main.js:16]: If we did that there would basically be nothing at all left in MainController, I think.  I don't have a strong opinion except that that kind of refactoring seems out of scope here. So I'll leave this to you or Justin to do in a separate pass if you want to.

File: app/js/views/main.js (hunk 12 above)
Comment (davidflanagan) [inline app/js/views/main.js:217]: I tried something like that, and had trouble getting the animation right. I could try again. A better approach might actually be to not even create those views until we open up, so we can keep the customizer really light-weight when it is not in use.  Again, though, that's out of scope for this bug.
